_posts/2018-11-08-google-cpp-style-guide-is-no-good.md

```diff
@@ -0,0 +1,297 @@
+---
+layout: post
+title: Google C++ Style Guide is No Good
+date: '2018-11-06T20:59:00.000-07:00'
+author: Eugene Yakubovich
+tags:
+- C++
+modified_time: '2018-10-23T20:59:46.484-07:00'
+---
+
+The [Google Style Guide (GSG)](https://google.github.io/styleguide/cppguide.html) for C++ has become popular outside of Google.
+But I don't like it at all. There... I said it.
+I want to take this opportunity to explain why I feel so strongly about it.
+
+## GSG "style"
+The general observation is that the style guide is written in a prohibitive fashion.
+Unlike C++ Core Guidelines that try to explain how to use the language effectively, GSG is about forbidding the use of certain features.
+Many of the rules prohibit the use of a feature over the fear of confusion, abuse, ambiguity and bugs.
+It is a well known fact that every feature (in every language) can be abused and misunderstood.
+Every line of code (and feature) can be a source of bugs.
+At the same time, every feature that went into the language was vetted for being useful.
+It was deemed that without the said feature, the code was significantly worse (e.g more verbose, prone to bugs).
+Therefore prohiting the use of some language feature just invites these problems into the codebase.
+Additionaly the GSG prohibits constructs that are widely used in well-known codebases: stdlib implementations, Boost, cppreference.com examples, etc.
+
+GSG is also concerned about some construct not being known to newcomers or even experienced C++ developers and outlaws such constructs based on these fears.
+But how will developers learn new features if not from reading code that uses them?
+Lastly, simple APIs often require complex implementations.
+Consider, for example, the following simple code snippet:
+
+{% highlight c++ linenos %}
+{% raw %}
+std::vector<int> v;
+v.resize(10, 42);
+auto it = v.begin();
+std::advance(it, 5);
+{% endraw %}
+{% endhighlight %}
+
+The above snippet invokes code paths that use constructs like `std::uninitialized_fill` (which uses placement new), `std::move_if_noexcept` (which uses SFINAE), implementation of random access iterator and tag dispatching.
```

Comment:
I assume these are things that the guide doesn't approve ?

Reply:
Some of them are explicitly forbidden. But the point was that these are all advanced features/techniques that may not be known to everyone. I'll make it clearer.